_posts/2020-11-09-defensive-programing.md

```diff
@@ -0,0 +1,110 @@
+---
+layout: post
+title:  "Defensive programming - what it is and why it is important"
+date:   2020-11-09 12:54:04 -0700
+categories: defensive programming
+---
+# **Defensive programming - what it is and why it is important**
+Gabrielly Lunkes  
+
+![Dwight Schrute from The Office show saying "HE PROTEC"]({{site.url }}https://philippegroarke.com/img/2018/defensive.png)
+
+Often times, a piece of code is used by different people, not only the person who created it. Also, codes can frequently be used to process different types of data. In my opinion, this flexibility to use the same code to process different datasets is one of the greatest things about coding. However, how is it possible to write a complex code in a way that other users would be able to easily navigate through and even use it with different types of input without bugs? Yes, you guessed it, the answer is in defensive programming.
+
+## What is defensive programming?
+
+In short, defensive programming is a way to make sure that your code works even when given incorrect input. This can be done by adding guard statements and assertions in your code that check if certain requirements are met and return a statement. The statements are usually helpful sentences that identify which requirement was not met and can even tell you what you should do to fix it. Here is an example:
+```r
+if (nrow(input_data) == 0) {
+  stop("ERROR: Input data is empty. Did you make a typo?")
+```
+
+In this case, the code is checking if the input file actually has data in it. If the file is empty, it will show the message `ERROR: Input data is empty. Did you make a typo?`. Notice that the statement points out what went wrong and gives an idea of how to solve the problem.
+
+## Some methods of defensive programming
+
+* **Assertions**
+
+Assertions are statements that check if a condition is true or false. If it is, the program does nothing. If the condition is false, the program will print an error message.
```

Comment:
You may have left out the word "true" in the line "If it is, the program does nothing." Also, you can perhaps explain how if the statement is true, the code will run all the way through and give some output.

Reply:
Yes, I did. I changed the sentence and somehow forgot to check it still worked well with the other sentences in the paragraph. Thanks for pointing it out!